cql3/restrictions/statement_restrictions.cc

```diff
@@ -1458,8 +1447,180 @@ bool is_satisfied_by(
+        {&operator_type::EQ, &operator_type::LT, &operator_type::LTE}, // These mean an upper bound.
+    };
+    const auto zero_if_lower_one_if_upper = get_idx(bnd);
+    return boost::algorithm::any_of_equal(operators[zero_if_lower_one_if_upper], op);
+}
+
+bound_t get_bound(const expression& restr, const query_options& options, statements::bound bnd) {
+    return std::visit(overloaded_functor{
+            [&] (bool v) {
+                if (v) {
+                    return bound_t(byte_type);
+                } else {
+                    throw std::logic_error("empty set has no bounds");
+                }
+            },
+            [&] (const conjunction& conj) {
+                auto invoke_get_bound = [&] (const expression& e) { return get_bound(e, options, bnd); };
+                auto pick_tighter = is_start(bnd) ? &tighter_lb : &tighter_ub;
+                // All conjunction elements have the same LHS; this is how
+                // single_column_restrictions::add_restriction constructs it.  We pick the tightest among all
+                // children's bounds.
+                return boost::accumulate(conj.children | transformed(invoke_get_bound),
+                                         bound_t(child_type(conj.children)), pick_tighter);
+            },
+            [&] (const binary_operator& opr) {
+                return std::visit(overloaded_functor{
+                        [&] (const std::vector<column_value>& cvs) {
+                            if (cvs.size() != 1) {
+                                throw std::logic_error("get_bound invoked on multi-column restriction");
+                            }
+                            auto cmptype = comparator(cvs[0]);
+                            return matches(opr.op, bnd) ?
+                                    bound_t(cmptype, to_bytes_opt(opr.rhs->bind_and_get(options)))
+                                    : bound_t(cmptype);
+                        },
+                        [&] (const token& tok) {
+                            return matches(opr.op, bnd) ?
+                                    bound_t(long_type, to_bytes_opt(opr.rhs->bind_and_get(options)))
+                                    : bound_t(long_type);
+                        },
+                    }, opr.lhs);
+            },
+        }, restr);
+}
+
+/// Finds the first multi-column binary_operator in restr that represents bnd and returns its RHS value.  If
+/// no such binary_operator exists, returns an empty vector.  The search is depth first.
+std::vector<bytes_opt> multicolumn_bound(const expression& restr, const query_options& options, statements::bound bnd) {
+    std::vector<bytes_opt> empty;
+    return std::visit(overloaded_functor{
+            [&] (const conjunction& conj) {
+                for (const auto& c : conj.children) {
+                    auto cb = multicolumn_bound(c, options, bnd);
+                    if (!cb.empty()) {
+                        return cb;
```

Comment:
Please add a comment explaining why we're so sure that there's only one multicolumn bound.

Reply:
Not sure what you mean by "only one".  If there are multiple bounds, we'll find the tightest one, same as `get_bound()`.  I could change the `cb.empty()` check to `cb==empty`, if that helps?

Reply:
I don't understand how you find the tightest bound. It looks like you're iterating over all children, and as soon as one is a multicolumn bound, you return.

Reply:
Oh, sorry, you're right.  We don't actually find the tightest bound, we find the first one as the function blurb suggests.  The reason it works is because upstream validation doesn't allow multiple multi-column atoms with same bound type:
```cql
cqlsh:ks> SELECT * FROM tbl WHERE p=1 AND (c1,c2)>(0,0) AND (c1,c2)>=(0,0);
InvalidRequest: Error from server: code=2200 [Invalid query] message="More than one restriction was found for the start bound on c1 ,c2"
cqlsh:ks> SELECT * FROM tbl WHERE p=1 AND (c1,c2)>(0,0) AND (c1,c2)=(0,0);
InvalidRequest: Error from server: code=2200 [Invalid query] message="Column "c1 ,c2" cannot be restricted by both an equality and an inequality relation"
```
To reply to your original comment: is the function blurb not enough?  It states explicitly that we find the first multi-column bound if it exists.

Reply:
Sure it says it finds the first, but doesn't explain why the first is sufficient. The data structure here allows multiple bounds, so it looks like a bug, even though it is blocked upstream.

Reply:
I think I'll rename this function to first_multicolumn_bound.  It doesn't need to explain why first is correct, because "correct" is in the eyes of the caller.  The caller should explain why it picks the first bound and not something else.

It is also possible that this will change in subsequent steps; the current setup is imposed by the old-code boundary I chose to stop at.

Reply:
Renamed as suggested above.